### PR TITLE
feat: [CDS-57982]: add API for validating template inputs for service yaml

### PR DESCRIPTION
**Pull request by hingerabhinav**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45575)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java [hunk 1]
```diff
@@ -884,6 +885,22 @@ public ResponseDTO<ServiceInputsMergedResponseDto> mergeServiceInputs(
         format("Service with type: [%s] does not support service hooks", serviceSpecType));
   }
 
+  @GET
+  @Path("validate-template-inputs")
+  @ApiOperation(value = "This validates whether yaml of service is valid or not", nickname = "validateTemplateInputs")
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java:890]: can we make the description more information to include artifact source templates ?

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java [hunk 2]
```diff
@@ -1003,4 +1009,22 @@ public Optional<ServiceEntity> getService(
...
+  @Override
+  public ValidateTemplateInputsResponseDTO validateTemplateInputs(
+      String accountId, String orgId, String projectId, String serviceIdentifier) {
+    Optional<ServiceEntity> optionalService = get(accountId, orgId, projectId, serviceIdentifier, false);
+
+    if (optionalService.isPresent()) {
+      String yaml = optionalService.get().getYaml();
+      RefreshRequestDTO refreshRequest = RefreshRequestDTO.builder().yaml(yaml).build();
+
+      if (TemplateRefHelper.hasTemplateRef(yaml)) {
```
Comment (yogesh-chauhan) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java:1022]: RefreshRequestDTO refreshRequest can be defined within this if block itself, its not used outside of this scope. Even better we can inline the object creation within the method argument

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java [hunk 3]
```diff
@@ -1003,4 +1009,22 @@ public Optional<ServiceEntity> getService(
...
         accountId, orgIdentifier, projectIdentifier, serviceIdentifier);
   }
+
+  @Override
+  public ValidateTemplateInputsResponseDTO validateTemplateInputs(
+      String accountId, String orgId, String projectId, String serviceIdentifier) {
+    Optional<ServiceEntity> optionalService = get(accountId, orgId, projectId, serviceIdentifier, false);
+
+    if (optionalService.isPresent()) {
+      String yaml = optionalService.get().getYaml();
```
Comment (yogesh-chauhan) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java:1019]: what if the yaml is empty ? I think older services might not have yaml
Either you can use io.harness.ng.core.service.entity.ServiceEntity.fetchNonEmptyYaml or check for empty

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java [hunk 4]
```diff
@@ -1003,4 +1009,22 @@ public Optional<ServiceEntity> getService(
     return serviceRepository.findByAccountIdAndOrgIdentifierAndProjectIdentifierAndIdentifier(
         accountId, orgIdentifier, projectIdentifier, serviceIdentifier);
   }
+
+  @Override
+  public ValidateTemplateInputsResponseDTO validateTemplateInputs(
+      String accountId, String orgId, String projectId, String serviceIdentifier) {
```
Comment (yogesh-chauhan) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java:1015]: is this method able to handle empty serviceId and other fields being empty or invalid with proper error message ?

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]
Please write unit test as well

**Review by hingerabhinav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45575)* status: 4 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java` line 890 at r1](https://reviewable.io/reviews/harness/harness-core/45575#-NREie9yPUu-4odjf9iv-r1-890:-NRfMaJG3JQX6xUcwgCE:bajublf) ([raw file](https://github.com/harness/harness-core/blob/9f3442dd3b41e2ba7e27f1bc452bf16eec050a64/120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java#L890)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we make the description more information to include artifact source templates ?
</blockquote></details>

Done. Wanted to keep it generic in case we have more templates in service. 
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java` line 1015 at r1](https://reviewable.io/reviews/harness/harness-core/45575#-NREie9yPUu-4odjf9iu-r1-1015:-NRfMgocESSEpPpu5o1M:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9f3442dd3b41e2ba7e27f1bc452bf16eec050a64/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java#L1015)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

is this method able to handle empty serviceId and other fields being empty or invalid with proper error message ?
</blockquote></details>

Done.
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java` line 1019 at r1](https://reviewable.io/reviews/harness/harness-core/45575#-NREie9yPUu-4odjf9iu-r1-1019:-NRfMhid1TBWHuILnlXn:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9f3442dd3b41e2ba7e27f1bc452bf16eec050a64/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java#L1019)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

what if the yaml is empty ? I think older services might not have yaml
Either you can use io.harness.ng.core.service.entity.ServiceEntity.fetchNonEmptyYaml or check for empty
</blockquote></details>

Done.
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java` line 1022 at r1](https://reviewable.io/reviews/harness/harness-core/45575#-NREie9yPUu-4odjf9iu-r1-1022:-NRfMiRX8mrPBlhgfGSD:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9f3442dd3b41e2ba7e27f1bc452bf16eec050a64/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java#L1022)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

RefreshRequestDTO refreshRequest can be defined within this if block itself, its not used outside of this scope. Even better we can inline the object creation within the method argument
</blockquote></details>

Done.
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java` line 1024 at r1](https://reviewable.io/reviews/harness/harness-core/45575#-NREie9yPUu-4odjf9iu-r1-1024:-NRfMiz56R04cWceRC8b:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9f3442dd3b41e2ba7e27f1bc452bf16eec050a64/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java#L1024)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we take loadFromCache as part of the API request ?
</blockquote></details>

Done.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45575)* status: 4 file(s) to review. 1 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java` line 890 at r1](https://reviewable.io/reviews/harness/harness-core/45575#-NRWTN8FA5Jir6Rc3E-J:-NRWTN8FA5Jir6Rc3E-K:b-feuhab) ([raw file](https://github.com/harness/harness-core/blob/9f3442dd3b41e2ba7e27f1bc452bf16eec050a64/120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java#L890)):*
> ```Java
>   @GET
>   @Path("validate-template-inputs")
>   @ApiOperation(value = "This validates whether yaml of service is valid or not", nickname = "validateTemplateInputs")
> ```

I think we should say that this applies only to artifact source templates for now to avoid confusion

**Comment by yogesh-chauhan** [comment]
Lets make sure to test with org/account scoped services

**Review by utkarshchoubey-harness — APPROVED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45575)* status: :shipit: complete! ✓

**Pull request merged**